File: lib/biz/spec-iot/src/main/java/site/addzero/biz/spec/iot/tdengine/TelemetryHistoryQuery.java
package site.addzero.biz.spec.iot.tdengine;

import site.addzero.biz.spec.iot.IotThingRef;

import java.time.LocalDateTime;

/**
 * History query input DTO.
 */
public final class TelemetryHistoryQuery {

    private final IotThingRef schemaRef;
    private final IotThingRef sourceRef;
    private final String identifier;
    private final LocalDateTime fromTime;
    private final LocalDateTime toTime;

    public TelemetryHistoryQuery(IotThingRef schemaRef, IotThingRef sourceRef, String identifier, LocalDateTime fromTime, LocalDateTime toTime) {
        if (schemaRef == null || sourceRef == null) {
            throw new IllegalArgumentException("schemaRef and sourceRef must not be null");
        }
        if (identifier == null || identifier.trim().isEmpty()) {
            throw new IllegalArgumentException("identifier must not be blank");
        }
        if (fromTime == null || toTime == null) {
            throw new IllegalArgumentException("fromTime and toTime must not be null");
        }
        this.schemaRef = schemaRef;
        this.sourceRef = sourceRef;
        this.identifier = identifier.trim();
        this.fromTime = fromTime;
        this.toTime = toTime;
    }

    public IotThingRef getSchemaRef() {
        return schemaRef;
    }

    public IotThingRef getSourceRef() {
        return sourceRef;
    }

    public String getIdentifier() {
        return identifier;
    }

    public LocalDateTime getFromTime() {
        return fromTime;
    }

    public LocalDateTime getToTime() {
        return toTime;
    }
}

File: lib/biz/spec-iot/src/main/kotlin/site/addzero/biz/spec/iot/tdengine/TelemetryHistoryQuery.kt
package site.addzero.biz.spec.iot.tdengine

import site.addzero.biz.spec.iot.IotThingRef
import site.addzero.biz.spec.iot.requireText
import java.time.LocalDateTime

/**
 * History query input DTO.
 */
class TelemetryHistoryQuery(
    schemaRef: IotThingRef?,
    sourceRef: IotThingRef?,
    identifier: String?,
    fromTime: LocalDateTime?,
    toTime: LocalDateTime?,
) {

    val schemaRef: IotThingRef = schemaRef ?: throw IllegalArgumentException("schemaRef and sourceRef must not be null")
    val sourceRef: IotThingRef = sourceRef ?: throw IllegalArgumentException("schemaRef and sourceRef must not be null")
    val identifier: String = requireText(identifier, "identifier")
    val fromTime: LocalDateTime = fromTime ?: throw IllegalArgumentException("fromTime and toTime must not be null")
    val toTime: LocalDateTime = toTime ?: throw IllegalArgumentException("fromTime and toTime must not be null")
}
